sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rule/BeamAggregateProjectMergeRule.java

```diff
@@ -0,0 +1,93 @@
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.RelNode;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.SingleRel;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Aggregate;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Filter;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.Project;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.core.RelFactories;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.rel.rules.AggregateProjectMergeRule;
+import org.apache.beam.vendor.calcite.v1_20_0.org.apache.calcite.tools.RelBuilderFactory;
+
+/**
+ * This rule is essentially a wrapper around Calcite's {@code AggregateProjectMergeRule}. In the
+ * case when an underlying IO supports project push-down it is more efficient to not merge {@code
+ * Project} with an {@code Aggregate}, leaving it for the {@code BeamIOPUshDownRule}.
+ */
+public class BeamAggregateProjectMergeRule extends AggregateProjectMergeRule {
+  public static final AggregateProjectMergeRule INSTANCE =
+      new BeamAggregateProjectMergeRule(
+          Aggregate.class, Project.class, RelFactories.LOGICAL_BUILDER);
+
+  public BeamAggregateProjectMergeRule(
+      Class<? extends Aggregate> aggregateClass,
+      Class<? extends Project> projectClass,
+      RelBuilderFactory relBuilderFactory) {
+    super(aggregateClass, projectClass, relBuilderFactory);
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final Project project = call.rel(1);
+    BeamIOSourceRel io = getUnderlyingIO(new HashSet<>(), project);
+
+    // Only perform AggregateProjectMergeRule when IO is not present or project push-down is not
+    // supported.
+    if (io == null || !io.getBeamSqlTable().supportsProjects().isSupported()) {
+      super.onMatch(call);
+    }
+  }
+
+  /**
+   * Following scenarios are possible:<br>
+   * 1) Aggregate <- Project <- IO.<br>
+   * 2) Aggregate <- Project <- Chain of Project/Filter <- IO.<br>
+   * 3) Aggregate <- Project <- Something else.<br>
+   * 4) Aggregate <- Project <- Chain of Project/Filter <- Something else.
+   *
+   * @param parent project that matched this rule.
+   * @return {@code BeamIOSourceRel} when it is present or null when some other {@code RelNode} is
+   *     present.
+   */
+  private BeamIOSourceRel getUnderlyingIO(Set<RelNode> visitedNodes, SingleRel parent) {
+    // No need to look at the same node more than once.
+    if (visitedNodes.contains(parent)) {
+      return null;
+    }
+    visitedNodes.add(parent);
+    List<RelNode> nodes = ((RelSubset) parent.getInput()).getRelList();
+
+    for (RelNode node : nodes) {
+      if (node instanceof Filter || node instanceof Project) {
+        return getUnderlyingIO(visitedNodes, (SingleRel) node);
```

Comment:
By returning here we may potentially skip some nodes, thus making this search incomprehensive.

Reply:
Added a check to return from that code path only when an IO is found.